### ISSUE TITLE
Scenario editor

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#558]
It would be super cool to have an in-game graphical scenario editor, where you could move a cursor around, draw terrain and entities, place robots, create custom entities and code, etc...

Of course the scenarios would still be stored in `.yaml` files, so creating them by hand will always be an option.

This would obviously be a massive undertaking, and I don't think it needs to happen any time soon, but I just thought of it today so writing it down!

**Comment by xsebek** [comment]
Selecting a cell with right click and then drawing by holding left mouse button would be pretty sweet. 🎨

**Comment by xsebek** [comment]
For inspiration, [asciiflow](https://asciiflow.com/) has a simple editor with some line drawing logic that could be useful for mazes.

**Comment by byorgey** [comment]
Wait, I just had an epiphany --- we don't need to make a scenario editor, because you should be able to just *program your own*!  A few ingredients would go into this:

- #102 would let you move a robot around with arrow keys, program your own hotkeys for placing certain items, drawing certain structures, etc., and of course you could always 
- We should add a built-in scenario that starts with a *blank* world
- There needs to be a way to save the current world in the format expected by a scenario description (related to #50).

**Comment by byorgey** [comment]
Well, OK, being able to draw stuff with the mouse would still be pretty cool, too.  But maybe we can think of an in-game way to do that---like some device that lets you receive mouse events and respond to them somehow, similar to #102 .  If we also had the ability to teleport entities in creative mode, you could then write a little program that responds to mouse events by teleporting certain entities to the proper coordinates.

**Comment by byorgey** [comment]
The overarching principle here is that this is a *programming* game --- instead of building in more and more cool features into the game itself, we should just build in programming language features that let players build the cool game features they want. (It even already says this in [DESIGN.md](https://github.com/swarm-game/swarm/blob/main/DESIGN.md),  believe it or not. :smile: )

**Comment by xsebek** [comment]
That is quite mind blowing and I love it! 🤯 Making mouse events an upgrade sounds good to me, I guess you just get coordinates and a boolean. 🤔 

Making saves compatible with scenarios would be great and this interactive workflow could really save time.

**Comment by byorgey** [comment]
> Making mouse events an upgrade sounds good to me, I guess you just get coordinates and a boolean.

I guess it depends how sophisticated you want to be.  Many real GUI frameworks differentiate between more than two kinds of mouse events, e.g. mouse down, mouse up, mouse move, etc.

**Comment by xsebek** [comment]
We still might need a form to set some parts of the game like the world function. After you are done with the scenario you would probably want to e.g. reset time, add win conditions and description.

**Comment by byorgey** [comment]
Yes, that's very true.  And maybe it's worth building some of that into the app so that you can create a scenario without directly editing a `.yaml` file.  But at least we don't need to build an actual graphical world editor.

**Comment by byorgey** [comment]
Note, we do now have a prototype in-game editor, though it's still somewhat limited.
